File: src-tauri/src/commands/sessions.rs
/**
 * Tauri commands for session management
 */

use crate::db::user::open_user_db;
use crate::services::sessions::{delete_session, get_all_sessions, get_session, get_sessions_by_language, get_session_words, SessionData, SessionWord};

/// Get all sessions (all languages)
#[tauri::command]
pub async fn get_all_sessions_command(app_handle: tauri::AppHandle) -> Result<Vec<SessionData>, String> {
    let pool = open_user_db(&app_handle).await.map_err(|e| e.to_string())?;
    get_all_sessions(&pool)
        .await
        .map_err(|e| e.to_string())
}

/// Get a single session by ID
#[tauri::command]
#[allow(non_snake_case)]
pub async fn get_session_command(app_handle: tauri::AppHandle, sessionId: String) -> Result<SessionData, String> {
    let pool = open_user_db(&app_handle).await.map_err(|e| e.to_string())?;
    get_session(&pool, &sessionId)
        .await
        .map_err(|e| e.to_string())
}

/// Get sessions filtered by language
#[tauri::command]
pub async fn get_sessions_by_language_command(app_handle: tauri::AppHandle, language: String) -> Result<Vec<SessionData>, String> {
    let pool = open_user_db(&app_handle).await.map_err(|e| e.to_string())?;
    get_sessions_by_language(&pool, &language)
        .await
        .map_err(|e| e.to_string())
}

/// Get vocabulary words for a session
#[tauri::command]
#[allow(non_snake_case)]
pub async fn get_session_words_command(app_handle: tauri::AppHandle, sessionId: String) -> Result<Vec<SessionWord>, String> {
    let pool = open_user_db(&app_handle).await.map_err(|e| e.to_string())?;
    get_session_words(&pool, &sessionId)
        .await
        .map_err(|e| e.to_string())
}

/// Delete a session and its related data
#[tauri::command]
#[allow(non_snake_case)]
pub async fn delete_session_command(app_handle: tauri::AppHandle, sessionId: String) -> Result<(), String> {
    println!("[delete_session_command] Received request to delete session: {}", sessionId);
    let pool = open_user_db(&app_handle).await.map_err(|e| {
        println!("[delete_session_command] Failed to open DB: {}", e);
        e.to_string()
    })?;
    println!("[delete_session_command] DB opened, calling delete_session service");
    delete_session(&pool, &sessionId)
        .await
        .map_err(|e| {
            println!("[delete_session_command] Delete failed with error: {}", e);
            e.to_string()
        })?;
    println!("[delete_session_command] Delete completed successfully");
    Ok(())
}

File: src/services/sessions/sessions.ts
/**
 * Sessions service - wraps Tauri commands for session management
 */

import { invoke } from '@tauri-apps/api/core';
import type { SessionData, SessionWord } from './types';
import { logger } from '@/services/logger'

/**
 * Get all sessions (all languages)
 */
export async function getAllSessions(): Promise<{ success: boolean; data?: SessionData[]; error?: string }> {
  try {
    const sessions = await invoke<SessionData[]>('get_all_sessions_command');
    return { success: true, data: sessions };
  } catch (error) {
    console.error('Failed to get all sessions:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Get a single session by ID
 */
export async function getSession(sessionId: string): Promise<{ success: boolean; data?: SessionData; error?: string }> {
  try {
    logger.debug('Fetching session with ID:', sessionId);
    const session = await invoke<SessionData>('get_session_command', { sessionId });
    logger.debug('Session fetched successfully', undefined, session);
    return { success: true, data: session };
  } catch (error) {
    console.error('Failed to get session:', error);
    console.error('Session ID was:', sessionId);
    return { success: false, error: String(error) };
  }
}

/**
 * Get sessions filtered by language
 */
export async function getSessionsByLanguage(language: string): Promise<{ success: boolean; data?: SessionData[]; error?: string }> {
  try {
    const sessions = await invoke<SessionData[]>('get_sessions_by_language_command', { language });
    return { success: true, data: sessions };
  } catch (error) {
    console.error('Failed to get sessions by language:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Get vocabulary words for a session
 */
export async function getSessionWords(sessionId: string): Promise<{ success: boolean; data?: SessionWord[]; error?: string }> {
  try {
    const words = await invoke<SessionWord[]>('get_session_words_command', { sessionId });
    return { success: true, data: words };
  } catch (error) {
    console.error('Failed to get session words:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Delete a session and its related data
 */
export async function deleteSession(sessionId: string): Promise<{ success: boolean; error?: string }> {
  try {
    logger.debug('[deleteSession] Deleting session:', sessionId);
    await invoke('delete_session_command', { sessionId });
    logger.debug('Session deleted successfully', 'deleteSession');
    return { success: true };
  } catch (error) {
    logger.error('Failed to delete session:', 'deleteSession', error);
    return { success: false, error: String(error) };
  }
}

export const sessionsService = {
  getAllSessions,
  getSession,
  getSessionsByLanguage,
  getSessionWords,
  deleteSession,
};
